File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.java
class Solution {
    public int shortestPathBinaryMatrix(int[][] grid) {
        if(grid[0][0] == 1) return -1;
        int n = grid.length;
        boolean[][] visited = new boolean[n][n];
        Queue<Point> q = new LinkedList<>();
        int[][] directions = new int[][]{{-1,-1},{-1,0},{-1,1},{0,-1},{0,1},{1,-1},{1,0},{1,1}};
        visited[0][0] = true;
        q.add(new Point(0,0,1));
        int count = 0;
        while(!q.isEmpty()) {
            Point p = q.poll();

            if(p.x == n - 1 && p.y == n - 1) return p.level;
            for(int[] dir : directions) {
                int newRow = p.x - dir[0];
                int newCol = p.y - dir[1];
                if(newRow < 0 || newRow == n || newCol < 0 || newCol == n || grid[newRow][newCol] == 1 || visited[newRow][newCol]) continue;
                //System.out.println(newRow + " - " + newCol);

                visited[newRow][newCol] = true;
                q.add(new Point(newRow,newCol,p.level + 1));
            }
        }
        
        return -1;
    }
}

class Point {
    int x;
    int y;
    int level;
    
    Point(int _x,int _y,int _level) {
        x = _x;
        y = _y;
        level = _level;
    }
}

File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestPathBinaryMatrix = function(grid) {
    
    if(grid[0][0] !== 0 || grid[grid.length-1][grid[0].length-1]) return -1;
    
    let getNeighbors = (cell) => {
        let directions = [[-1,-1],[-1,0],[-1,1],[0,-1],[0,1],[1,-1],[1,0],[1,1]];
        let neighbors = [];
        for (let dir of directions) {
            let newRow = dir[0] + cell[0];
            let newCol = dir[1] + cell[1];
            
            if(newRow < 0 || newRow > grid.length - 1 ||
              newCol < 0 || newCol > grid[0].length - 1 ||
              grid[newRow][newCol] !== 0)
                continue;
            
            neighbors.push([newRow,newCol]);
        }
        return neighbors;
    }
    
    let q =[];
    q.push([0,0]);
    grid[0][0] = 1;
    let distance = -1;
    while(q.length > 0){
        let cell = q.shift();
        distance = grid[cell[0]][cell[1]];
        
        if(cell[0] === grid.length - 1 && cell[1] === grid[0].length - 1)
            return distance;
        
        let neighborList = getNeighbors(cell);
        
        for( let n of neighborList) {
            grid[n[0]][n[1]] = distance + 1;
            q.push(n);
        }
    }
    return -1;
};